File: src/ai_multi_reference_timekeeping/metrics.py
"""Clock quality metrics for timekeeping evaluation."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Iterable, Sequence


@dataclass(frozen=True)
class HoldoverStats:
    """Holdover behavior summary."""

    max_offset: float
    rms_offset: float
    duration: float


def tdev(offsets: Sequence[float], tau: int) -> float:
    """Compute a simple TDEV estimate for a given tau (samples)."""

    if tau <= 0:
        raise ValueError("tau must be positive")
    if len(offsets) < 3 * tau:
        raise ValueError("Not enough samples for TDEV")
    diffs = []
    for i in range(len(offsets) - 3 * tau + 1):
        avg1 = sum(offsets[i : i + tau]) / tau
        avg2 = sum(offsets[i + tau : i + 2 * tau]) / tau
        avg3 = sum(offsets[i + 2 * tau : i + 3 * tau]) / tau
        diffs.append(avg1 - 2 * avg2 + avg3)
    variance = sum(diff * diff for diff in diffs) / (2.0 * len(diffs))
    return variance**0.5


def mtie(offsets: Sequence[float], window: int) -> float:
    """Compute MTIE for a sliding window of length `window`."""

    if window <= 0:
        raise ValueError("window must be positive")
    if len(offsets) < window:
        raise ValueError("Not enough samples for MTIE")
    max_tie = 0.0
    for i in range(len(offsets) - window + 1):
        segment = offsets[i : i + window]
        tie = max(segment) - min(segment)
        if tie > max_tie:
            max_tie = tie
    return max_tie


def holdover_stats(offsets: Iterable[float], sample_interval: float) -> HoldoverStats:
    """Compute holdover stats over a sequence of offsets."""

    if sample_interval <= 0:
        raise ValueError("sample_interval must be positive")
    values = list(offsets)
    if not values:
        raise ValueError("offsets must be non-empty")
    max_offset = max(abs(value) for value in values)
    rms_offset = (sum(value * value for value in values) / len(values)) ** 0.5
    duration = sample_interval * (len(values) - 1)
    return HoldoverStats(max_offset=max_offset, rms_offset=rms_offset, duration=duration)

File: cpp/src/metrics.cpp
#include "aimrt/metrics.hpp"

#include <cmath>
#include <stdexcept>

namespace aimrt {

double tdev(const std::vector<double>& offsets, int tau) {
    if (tau <= 0) {
        throw std::runtime_error("tau must be positive");
    }
    if (offsets.size() < static_cast<size_t>(3 * tau)) {
        throw std::runtime_error("Not enough samples for TDEV");
    }

    std::vector<double> diffs;
    diffs.reserve(offsets.size() - static_cast<size_t>(3 * tau) + 1);
    for (size_t i = 0; i + static_cast<size_t>(3 * tau) <= offsets.size(); ++i) {
        double avg1 = 0.0;
        double avg2 = 0.0;
        double avg3 = 0.0;
        for (int j = 0; j < tau; ++j) {
            avg1 += offsets[i + j];
            avg2 += offsets[i + static_cast<size_t>(tau) + j];
            avg3 += offsets[i + static_cast<size_t>(2 * tau) + j];
        }
        avg1 /= tau;
        avg2 /= tau;
        avg3 /= tau;
        const double diff = avg1 - 2.0 * avg2 + avg3;
        diffs.push_back(diff);
    }

    double variance = 0.0;
    for (double diff : diffs) {
        variance += diff * diff;
    }
    variance /= (2.0 * diffs.size());
    return std::sqrt(variance);
}

double mtie(const std::vector<double>& offsets, int window) {
    if (window <= 0) {
        throw std::runtime_error("window must be positive");
    }
    if (offsets.size() < static_cast<size_t>(window)) {
        throw std::runtime_error("Not enough samples for MTIE");
    }

    double max_tie = 0.0;
    for (size_t i = 0; i + static_cast<size_t>(window) <= offsets.size(); ++i) {
        double min_value = offsets[i];
        double max_value = offsets[i];
        for (int j = 1; j < window; ++j) {
            const double value = offsets[i + static_cast<size_t>(j)];
            min_value = std::min(min_value, value);
            max_value = std::max(max_value, value);
        }
        max_tie = std::max(max_tie, max_value - min_value);
    }
    return max_tie;
}

HoldoverStats holdover_stats(const std::vector<double>& offsets, double sample_interval) {
    if (sample_interval <= 0.0) {
        throw std::runtime_error("sample_interval must be positive");
    }
    if (offsets.empty()) {
        throw std::runtime_error("offsets must be non-empty");
    }

    double max_offset = 0.0;
    double sum_sq = 0.0;
    for (double value : offsets) {
        max_offset = std::max(max_offset, std::abs(value));
        sum_sq += value * value;
    }

    const double rms_offset = std::sqrt(sum_sq / offsets.size());
    const double duration = sample_interval * static_cast<double>(offsets.size() - 1);
    return HoldoverStats{max_offset, rms_offset, duration};
}

}  // namespace aimrt
